docs/adr/0007-validation-guideliness.md

```diff
@@ -0,0 +1,146 @@
+# How we are going to validate the new API codebase
+
+* Date: 2021-03-10
+
+Technical Story:
+- [securesystemslib schema checker issues](https://github.com/secure-systems-lab/securesystemslib/issues/183)
+- [new TUF validation guidelines](https://github.com/theupdateframework/tuf/issues/1130)
+
+## Context and Problem Statement
+
+1. Some schemas sound more specific than they are.
+2. Some schemas are an odd replacement for constants.
+3. Schema validation is generally **overused**. Together with user input,
+we are validating input programmatically generated from our private functions.
+4. There are instances where some attributes are validated multiple times
+when executing one API call.
+5. Schema checking sometimes makes execution branches unreachable.
+6. The error messages from checking schemas are often not helpful.
+
+## Decision Drivers and Requirements
+Some of the requirements we want to meet are:
+1. The ability to decide which functions to validate and which not.
+2. Allow for custom deeper validation beyond type check.
+3. As little as possible performance overhead.
+4. Add as minimal number of dependencies as possible.
+5. Support for all python versions we are using.
+
+## Considered Options
+1. Usage of a `ValidationMixin`.
+2. Usage of a third-party library called `pydantic`.
+
+## Pros, Cons, and Considerations of the Options
+
+### Option 1: Usage of a ValidationMixin
+
+**Note:** All pros, cons, and considerations are documented with the assumption
+we would implement the `ValidationMixin` the same way it is implemented in
+[in-toto](https://github.com/in-toto) until version 1.0.1 (the latest
+version at the time of writing.)
+
+* Good, because it's shorter by calling one function and validating
+multiple fields.
+
+* Good, because it allows reuse of the validation code through
+`securesystemslib.schemas` or another schema of our choice.
```

Comment:
I think you are focusing too much on `securesystemslib.schemas`. There is an unspoken consensus that we want to get rid of it (see https://github.com/secure-systems-lab/securesystemslib/issues/183). And I don't think that its use is a defining aspect of the `ValidationMixin`. IIRC we only used them in in-toto because it was already there and we lacked a comprehensive class model for all complex data types.

Take for instance the [`_validate_keys` method on the in-toto `Layout` class](https://github.com/in-toto/in-toto/blob/06598a7b8c2e4a7dcbac2de898c16aa0843948ad/in_toto/models/layout.py#L449-L451):

```python
  def _validate_keys(self):
    securesystemslib.formats.ANY_PUBKEY_DICT_SCHEMA.check_match(self.keys)

```
Now if we had a `PublicKey` class with its own validators -- something we intend to add as per [ADR4](https://github.com/theupdateframework/tuf/blob/2c360d0538f73a59854add867234f48a47f7d9a3/docs/adr/0004-extent-of-OOP-in-metadata-model.md) particularly for the purpose of  validation -- then the validator would probably look like this:


```python
  def _validate_keys(self):
    if not isinstance(self.keys, dict):
      raise ...

    for keyid, key in self.keys.items():
      if not isinstance(key, PublicKey):
        raise ...
      key.validate()

      # NOTE: Even though ADR4 only talks about classes for complex attributes
      # it probably makes sense to add a `KeyID` class as well just so that we
      # have a reusable way of checking hex strings of a certain length.
      if not isinstance(keyid, KeyID):
        raise ...
      keyid.validate()
```

Or we could add a `PublicKeyDict` class, and do:

```python
  def _validate_keys(self):
    if not isinstance(self.keys, PublicKeyDict):
      raise ...
    self.keys.validate()
```

Or, and that's my preferred way, we let a type annotation checker do the basic type checking for us, so that we always know that an in-toto layout's `pubkeys` variable contains a valid `Dict[KeyID, PublicKey]` value. Then we can use the `_validate_*` for more complex checks, such as:
```python
  def _validate_keys(self):
    # We already know that 'self.pubkeys' is a valid Dict[KeyID, PublicKey]
    # value so no need to check this here...

    # ... but we also want to make sure that we don't have duplicate keys.
    if len(self.pubkeys.keys()) != len(set(self.pubkeys.keys())):
      raise ...
```

I'd say the `ValidationMixin` really just provides a shortcut -- i.e. `validate()` --  to all `_validate_*` methods on a class, which seems very similar to pydantic's `@validator` decorator (please correct me if I'm wrong!!)

So the main question to me is, is that decorator better than our custom `_validate_*` convention, and does `pydantic` have other features (that we can't just easily add to our `ValidationMixin`) that justify a +7K LOC dependency. :)

Reply:
Agreed: we should use ideally no 3rd-party dependency.

Reply:
I tried to update the ADR to not sound like the idea with `ValidationMixin` fully depended on the schemas, but
it seems it didn't sound that way :D.

I understand the idea with the `ValidationMixin` and it has good points.
If we decide to go into that path, probably it's fixing the existing schemas instead of straightly removing them.

> Or, and that's my preferred way, we let a type annotation checker do the basic type checking for us,

@lukpueh which type annotation checker do you use in `in-toto`?
How many dependencies will it add into `tuf` and (if you can easily check that) how many lines of code? 

As I said in my detailed comment here https://github.com/theupdateframework/tuf/pull/1301#issuecomment-802043432 there are more useful features in `pydantic` which could be useful for us.